test(transformer/using): failing test for class name

overlookmotel opened the pull request:
Add a failing test for `using` transform where exported class has a reference to itself within the class body, and the external class binding is mutated.

This is something of an edge case, but it'd be ideal to handle it correctly.

Babel has the same bug: https://github.com/babel/babel/issues/17172

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9678?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9678** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9678?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9678?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9677** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9677?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9676** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9676?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9675** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9675?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9674** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9674?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9673** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9673?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9672** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9672?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9678).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 reviewed (approved):
good find! - out of interest, how did you notices this issue, was it just from reading code, or testing or something else?

overlookmotel commented:
> good find! - out of interest, how did you notices this issue, was it just from reading code, or testing or something else?

Reading the code. But I'm familiar with these edge cases from working on the class properties transform.

We've found during past months on transformer that Babel's test cases often don't cover as much as you'd like. Sometimes it's just edge case bugs which are rare in practice, so no-one has noticed. But also some things are really simple to do in JS, so Babel probably didn't see the need to add tests for things which are obvious - but they're not nearly as simple or obvious in Rust. So we've had to add quite a lot of extra tests on some transforms.

overlookmotel committed:
test(transformer/using): failing test for class name (#9678)

Add a failing test for `using` transform where exported class has a reference to itself within the class body, and the external class binding is mutated.

This is something of an edge case, but it'd be ideal to handle it correctly.

Babel has the same bug: https://github.com/babel/babel/issues/17172
